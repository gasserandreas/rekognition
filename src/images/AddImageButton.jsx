import React, { Fragment, useState, useRef } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import uuid from 'uuid';
import styled from 'styled-components';
import Dropzone from 'react-dropzone';

import { Add } from 'grommet-icons';

import AddImageMessage from './AddImageMessage';

import Button from '../ui/form/Button';

import * as reduxImages from '../redux/images';
import { addImageIsLoading } from '../redux/images/selectors';

import { Colors, MediaSize } from '../styles';

const maxFileSize = 6000000;

const StyledAddImageButton = styled(Button)`
  border-radius: 50%;
  width: 44px;
  height: 46px;
  position: fixed;
  right: 3rem;
  bottom: 2rem;
  box-shadow: 0px 8px 16px rgba(0,0,0,0.20);
  z-index: 100;

  @media (min-width: ${MediaSize.Notebook}) {
    right: 2.5rem;
    bottom: 2.5rem;
  }
`;

const initialState = {
  showMessage: false,
  image: null,
};

const AddImageButton = ({
  loading,
  addImage,
  afterOnClick,
  ...props
}) => {
  const [state, setState] = useState(initialState);
  const uploadRef = useRef();

  const handleResetForm = () => {
    setState({
      showMessage: false,
      image: null,
    });
  };

  const handleOpenImageDialog = () => {
    const dialog = uploadRef.current;

    if (dialog) {
      dialog.open();
    }
  };

  /* instanbul ignore-next */
  const handleUploadImage = (files) => {
    /* instanbul ignore-next */
    if (files.length > 0) {
      /* instanbul ignore-next */
      files.forEach((file) => {
        const { size } = file;
        if (size > maxFileSize) {
          // image is to big
          setState({
            showMessage: true,
            image: file,
          });
        } else {
          // upload
          /* instanbul ignore-next */
          const id = uuid.v4();
          /* instanbul ignore-next */
          addImage({ file, id });

          // reset form
          /* instanbul ignore-next */
          handleResetForm();

          /* instanbul ignore-next */
          if (afterOnClick) {
            /* instanbul ignore-next */
            afterOnClick();
          }
        }
      });
    }
  };

  const { showMessage, image } = state;
  return (
    <Fragment>
      <AddImageMessage
        show={showMessage}
        image={image}
        onHandleResetForm={handleResetForm}
        onHandleOpenImageDialog={handleOpenImageDialog}
      />
      <Dropzone
        onDrop={handleUploadImage}
        ref={uploadRef}
      >
        {({ getRootProps, getInputProps }) => (
          <span {...getRootProps()}>
            <input {...getInputProps()} />
            <StyledAddImageButton
              {...props}
              type="button"
              icon={<Add color={Colors.ColorsPalette.White} />}
              size="xlarge"
              buttonStyle="primary"
              elevation="medium"
              disabled={loading}
            />
          </span>
        )}
      </Dropzone>
    </Fragment>
  );
};

AddImageButton.propTypes = {
  loading: PropTypes.bool.isRequired,
  addImage: PropTypes.func.isRequired,
  afterOnClick: PropTypes.func,
};

AddImageButton.defaultProps = {
  afterOnClick: null,
};

// redux
const select = state => ({
  loading: addImageIsLoading(state),
});

const mapDispatchToProps = ({
  addImage: reduxImages.addImage,
});

export const __testables__ = {
  select,
  mapDispatchToProps,
  AddImageButton,
  StyledAddImageButton,
  maxFileSize,
};

export default connect(select, mapDispatchToProps)(AddImageButton);
