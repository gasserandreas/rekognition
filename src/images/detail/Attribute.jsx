import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';

import { Box } from 'grommet';

export const AttributePropType = PropTypes.shape({
  name: PropTypes.string.isRequired,
  confidence: PropTypes.number,
  value: PropTypes.string,
});

// components
const StyledConfidence = styled.span`
  display: inline-block;
  margin-left: 0.25rem;
  &:before {
    content: "(";
  }
  
  &:after {
    content: ")";
  }
`;

export const StyledAttrLabel = styled.label`
  font-weight: ${props => props.bold ? 600 : 400};
  flex-shrink: 0;
  flex-grow: 0;
  width: 7rem;
  margin-right: 0.5rem;
  text-transform: capitalize;
`;

export const StyledAttrContent = styled.div`
  flex-shrink: 1;
  flex-grow: 1;
`;

const StyledAttr = styled(Box)`
  margin-bottom: 0.25rem;
`;

export const Attribute = ({
  attribute: {
    name,
    confidence,
    value,
  },
  showConfidence,
  boldLabel,
}) => (
  <StyledAttr
    fill
    direction="row"
    alignContent="between"
  >
    <StyledAttrLabel bold={boldLabel}>{name}</StyledAttrLabel>
    <StyledAttrContent>
      {value}
      {showConfidence && (
        <StyledConfidence>{String(confidence).substring(0,5)}</StyledConfidence>
      )}
    </StyledAttrContent>
  </StyledAttr>
)

Attribute.propTypes = {
  attribute: AttributePropType.isRequired,
  showConfidence: PropTypes.bool,
  boldLabel: PropTypes.bool,
}

Attribute.defaultProps = {
  showConfidence: false,
  boldLabel: true,
};

export default Attribute;
