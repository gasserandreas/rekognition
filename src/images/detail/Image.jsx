import React, { useState, useEffect, useRef } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';

import { Box } from 'grommet';
import AsyncImage from '../../ui/async/AsyncImage';

import { getImageSrc } from '../../util/util';
import { Colors } from '../../styles';

// util
const generateInitPos = () => ({
  top: 0,
  left: 0,
  width: 0,
  height: 0,
});

// label / face selector
const StyledSelector = styled.div`
  border: 2px solid ${Colors.ColorsPalette.Brand};
  border-radius: 3px;
  margin: 0;
  padding: 0;
  position: absolute;
  top: ${props => props.pos.top * 100}%;
  left: ${props => props.pos.left * 100}%;
  height: ${props => props.pos.height * 100}%;
  width: ${props => props.pos.width * 100}%;
  z-index: 10;
`;

// image
const StyledAsyncImage = styled(AsyncImage)`
  width: 100%;
  height: 100%;
  max-width: 100%;
  max-height: 100%;
`;

// image wrapper
const StyledImageWrapper = styled.div`
  position: absolute;
  top: ${props => props.pos.top}px;
  left: ${props => props.pos.left}px;
  width: ${props => props.pos.width}px;
  height: ${props => props.pos.height}px;
`;

StyledImageWrapper.defaultProps = {
  pos: PropTypes.shape({
    top: 0,
    left: 0,
    width: 0,
    height: 0,
  }),
};

// image wrapper comp
const StyledImageContainer = styled(Box)``;

const getPositions = (meta, wrapperBounding) => {
  const { height: imageHeight, width: imageWidth, orientation } = meta;

  // get wrapper positions
  const imageContainerPosition = {
    x: wrapperBounding.x,
    y: wrapperBounding.y,
    width: wrapperBounding.width,
    height: wrapperBounding.height,
    top: wrapperBounding.top,
    bottom: wrapperBounding.bottom,
  };

  let width;
  let height;
  let ratio;

  if (orientation === 'LANDSCAPE') {
    // simple case, use full wrapper width
    width = imageContainerPosition.width; // eslint-disable-line prefer-destructuring

    // get height by sticking with the ratio
    ratio = imageWidth / width;
    height = imageHeight / ratio;

    // check if image height fits into viewport by using max width
    if (height > imageContainerPosition.height) {
      height = imageContainerPosition.height; // eslint-disable-line prefer-destructuring
      ratio = imageHeight / height;
      width = imageWidth / ratio;
    }
  } else {
    // portrait mode full height
    height = imageContainerPosition.height; // eslint-disable-line prefer-destructuring
    ratio = imageHeight / height;
    width = imageWidth / ratio;

    // check if image width fits into viewport by using max height
    if (width > imageContainerPosition.width) {
      // scale image down to full width
      width = imageContainerPosition.width; // eslint-disable-line prefer-destructuring
      ratio = imageWidth / width;
      height = imageHeight / ratio;
    }
  }

  // calculate top position (center of container)
  const top = imageContainerPosition.height / 2 - height / 2;
  const left = imageContainerPosition.width / 2 - width / 2;

  // define image position
  const imageWrapperPosition = {
    top,
    left,
    width,
    height,
  };

  return {
    imageContainerPosition,
    imageWrapperPosition,
  };
};

const ImageContainer = ({ image: { meta, path }, selectedFace, selectedLabel }) => {
  const wrapperRef = useRef();
  const [state, setState] = useState({
    imageWrapperPosition: generateInitPos(),
    imageContainerPosition: generateInitPos(),
  });

  useEffect(() => {
    const handleWrapperPosition = () => {
      const wrapperBounding = wrapperRef.current.getBoundingClientRect();

      // (deep) equals check based on string
      if (JSON.stringify(state.wrapperBounding) !== JSON.stringify(wrapperBounding)) {
        const { imageContainerPosition, imageWrapperPosition } = getPositions(meta, wrapperBounding);

        setState({
          imageContainerPosition,
          imageWrapperPosition,
        });
      }
    };

    // add event listener
    window.addEventListener('resize', handleWrapperPosition);

    // calculate wrapper position
    handleWrapperPosition();

    return () => {
      window.removeEventListener('resize', handleWrapperPosition);
    };
  }, [meta, path, state.wrapperBounding]);

  const { imageWrapperPosition } = state;

  return (
    <StyledImageContainer ref={wrapperRef} style={{ justifyContent: 'center' }} fill>
      <StyledImageWrapper pos={imageWrapperPosition}>
        {selectedFace && <StyledSelector id="jestSelectedFace" pos={selectedFace.position} />}
        {selectedLabel && (
          <span id="jestSelectedLabel">
            {selectedLabel.instances.map((pos, i) => (
              <StyledSelector key={`label_selector_${i}`} pos={pos} /> // eslint-disable-line react/no-array-index-key
            ))}
          </span>
        )}
        <StyledAsyncImage
          src={getImageSrc(path)}
          fit="contain"
          wrapperStyles={`
            position: relative;
            display: inline;
          `}
        />
      </StyledImageWrapper>
    </StyledImageContainer>
  );
};

ImageContainer.propTypes = {
  image: PropTypes.shape({
    meta: PropTypes.shape({}),
    path: PropTypes.string.isRequired,
  }).isRequired,
  selectedFace: PropTypes.shape({}),
  selectedLabel: PropTypes.shape({}),
};

ImageContainer.defaultProps = {
  selectedFace: null,
  selectedLabel: null,
};

export const __testables__ = {
  StyledSelector,
  StyledAsyncImage,
  StyledImageWrapper,
  generateInitPos,
  getPositions,
};

export default ImageContainer;
