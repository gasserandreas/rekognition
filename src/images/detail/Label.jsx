import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';

import { Colors } from '../../styles';

const hoverSelectedStyles = `
  cursor: pointer;
  background-color: ${Colors.Blue.Default};
  color: ${Colors.ColorsPalette.White};
`;

const handleLabelHover = ({ isClickable, selected }) => {
  if (selected) {
    return hoverSelectedStyles;
  }

  if (isClickable) {
    return `
      color: ${Colors.Blue.Default};
      border-color: ${Colors.Blue.Default};

      &:hover {
        ${hoverSelectedStyles}
      }
    `;
  }

  return '';
};

const StyledLabel = styled.div`
  border: 1.5px solid ${Colors.Neutrals.Mid};
  padding: 0.25rem 0.5rem;
  border-radius: 3px;
  display: inline-block
  font-size: 0.8rem;
  margin-right: 0.25rem;
  margin-bottom: 0.5rem;

  span {
    font-size: 0.75rem;
  }

  ${props => handleLabelHover(props)}
`;

const Label = (props) => {
  const { label, ...rest } = props;
  const { name, confidence } = label;

  return (
    <StyledLabel {...rest}>
      {name}
      {' '}
      {confidence && (
      <span id="jestConfidence">
(
        {confidence.toFixed(2)}
)
      </span>
      )}
    </StyledLabel>
  );
};

Label.propTypes = {
  label: PropTypes.shape({
    id: PropTypes.string,
    name: PropTypes.string.isRequired,
    confidence: PropTypes.number,
    parents: PropTypes.arrayOf(PropTypes.string),
    instances: PropTypes.arrayOf(
      PropTypes.shape({
        height: PropTypes.number.isRequired,
        width: PropTypes.number.isRequired,
        left: PropTypes.number.isRequired,
        top: PropTypes.number.isRequired,
      }),
    ),
  }).isRequired,
  onClick: PropTypes.func,
  isClickable: PropTypes.bool,
  selected: PropTypes.bool,
};

Label.defaultProps = {
  onClick: null,
  isClickable: false,
  selected: false,
};

export const __testables__ = {
  StyledLabel,
  handleLabelHover,
};

export default Label;
