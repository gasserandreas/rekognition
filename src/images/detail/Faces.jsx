import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';

import { Box, Text, Paragraph } from 'grommet';

import Label from './Label';
import { Attribute, StyledAttrLabel, AttributePropType } from './Attribute';

import { Colors } from '../../styles';

// emotions
const StyledEmotions = styled.div`
  label {
    margin-bottom: 0.25rem;
  }
`;

const Emotions = ({ faceId, emotions, ...props }) => (
  <StyledEmotions {...props}>
    <StyledAttrLabel
      style={{ display: 'block'}}
      bold
    >Emotions</StyledAttrLabel>
      {emotions
        .sort((a, b) => a.confidence < b.confidence ? 1 : -1)
        .map(({ name, confidence }) => (
          <Label
            key={`face_${faceId}_emotion_label_${name}`}
            label={{ name, confidence }}
          />
        )
      )}
  </StyledEmotions>
);

Emotions.propTypes = {
  faceId: PropTypes.string.isRequired,
  emotions: PropTypes.arrayOf(AttributePropType).isRequired,
};

// face component
const StyledFace = styled(Box)`
  margin-bottom: 0.5rem;

  &:hover {
    background-color: ${Colors.Neutrals.LightMid};
    cursor: pointer;
  }
`;

StyledFace.propTypes = {
  selected: PropTypes.bool,
}

StyledFace.defaultProps = {
  selected: false,
}

const Face = ({ face, number, ...props }) => {
  const { id, age, attributes, emotions } = face;

  // generate age attribute
  const ageAttribute = {
    name: 'age',
    confidence: 100,
    value: `${age.low} - ${age.high}`,
  };

  // generate other attributes
  const filteredAttributes = attributes.filter(item => (
    item.name !== 'brightness' &&
    item.name !== 'sharpness'
  ));

  return (
    <StyledFace {...props} >
      <h3 level="1">
        Face {number}
        {props.selected && <Text color="brand" size="small">{' '}(selected)</Text>}
      </h3>
      <Box>
        <Attribute attribute={ageAttribute} />
        {filteredAttributes.map(attr => (
          <Attribute
            attribute={attr}
            key={`face_attribute_${id}_${attr.name}`}
            showConfidence
          />
        ))}
        <Emotions
          faceId={id}
          emotions={emotions}
        />
      </Box>
    </StyledFace>
  );
};

Face.propTypes = {
  face: PropTypes.shape({
    id: PropTypes.string.isRequired,
    position: PropTypes.shape({
      height: PropTypes.number.isRequired,
      width: PropTypes.number.isRequired,
      left: PropTypes.number.isRequired,
      top: PropTypes.number.isRequired,
    }),
    age: PropTypes.shape({
      low: PropTypes.number.isRequired,
      high: PropTypes.number.isRequired,
    }),
    emotions: PropTypes.arrayOf(AttributePropType).isRequired,
    attributes: PropTypes.arrayOf(AttributePropType).isRequired,
  }).isRequired,
  onClick: PropTypes.func.isRequired,
}

// faces wrapper component
const StyledFaces = styled(Box)``;

const Faces = ({ faces, onFaceClick, selectedFace, ...props }) => (
  <StyledFaces {...props}>
    <Paragraph size="small" margin="none">Hint: Click on faces to visualize selected face.</Paragraph>
    {faces.map((face, i) => (
      <Face
        number={i + 1}
        key={`face_item_${face.id}`}
        face={face}
        onClick={() => onFaceClick(face)}
        selected={selectedFace && selectedFace.id === face.id}
      />
    ))}
  </StyledFaces>
);

Faces.propTypes = {
  faces: PropTypes.arrayOf(PropTypes.shape({})).isRequired,
  selectedFace: PropTypes.shape({}),
  onFaceClick: PropTypes.func.isRequired,
};

Faces.defaultProps = {
  selectedFace: {
    id: null,
  },
}

export default Faces;
