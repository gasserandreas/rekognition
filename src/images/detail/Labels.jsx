import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';

import { Box } from 'grommet';

import Label from './Label';

// labels
const StyledLabels = styled(Box)``;

const Labels = ({
  selectedLabel,
  labels,
  onLabelClick,
}) => {

  const handleOnLabelClick = (label) => {
    if (isClickEnabled(label)) {
      onLabelClick(label)
    }
  };

  const isClickEnabled = label => label.instances.length > 0;

  return (
    <StyledLabels
      wrap
      direction="row"
    >
      {labels.map((label) => {
        const labelProps = {
          key: `image_label_${label.id}`,
          label,
          onClick: () => handleOnLabelClick(label),
          isClickable: isClickEnabled(label),
          selected: selectedLabel && selectedLabel.id === label.id,
        };

        return <Label {...labelProps} />;
      })}
    </StyledLabels>
  );
};

Labels.propTypes = {
  labels: PropTypes.arrayOf(PropTypes.shape({})).isRequired,
  selectedLabel: PropTypes.shape({}),
  onLabelClick: PropTypes.func.isRequired,
};

Labels.defaultProps = {
  selectedLabel: {
    id: '',
  },
};

export default Labels;
