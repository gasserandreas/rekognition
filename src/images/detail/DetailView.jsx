import React, { Component } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';

import { Box, Heading } from 'grommet';

import Labels from './Labels';
import Faces from './Faces';

import { Colors, MediaSize, Sizes } from '../../styles';
import { getUrl } from '../../util/services/networkUtils';

import View from '../../ui/View';
import AsyncImage from '../../ui/AsyncImage';
import LoadingIndicator from '../../ui/LoadingIndicator';

import { HOCRequestPropTypes } from '../../util/PropTypes';

const getImageSrc = path => `${getUrl('thumb')}/${path}`;

const StyledHeading = styled(Heading)`
  margin-bottom: 0.5rem;
`;

const StyledImageBox = styled(Box)`
  width: 100%;
  position: relative;

  @media (min-width: ${MediaSize.Tablet}) {
    padding-left: ${Sizes.LeftBar.width.Tablet};
  }

  @media (min-width: ${MediaSize.Notebook}) {
    padding-left: ${Sizes.LeftBar.width.Notebook};
  }

  @media (min-width: ${MediaSize.Desktop}) {
    padding-left: ${Sizes.LeftBar.width.Desktop};
  }

  @media (min-width: ${MediaSize.Fullscreen}) {
    padding-left: ${Sizes.LeftBar.width.Fullscreen};
  }
`;

const StyledDataBox = styled(Box)`
  width: 100%;
  background-color: ${Colors.Neutrals.Light};

  @media (min-width: ${MediaSize.Tablet}) {
    position: fixed;
    top: ${Sizes.Header.height};
    left: 0;
    bottom: 0;
    background-color: ${Colors.Neutrals.Light};
    box-shadow: 0px 2px 4px rgba(0,0,0,0.20);
    z-index: 51;

    width: ${Sizes.LeftBar.width.Tablet};
  }

  @media (min-width: ${MediaSize.Notebook}) {
    width: ${Sizes.LeftBar.width.Notebook};
  }

  @media (min-width: ${MediaSize.Desktop}) {
    width: ${Sizes.LeftBar.width.Desktop};
  }

  @media (min-width: ${MediaSize.Fullscreen}) {
    width: ${Sizes.LeftBar.width.Fullscreen};
  }
`;

const StyledScrollableData = styled.div`

  @media (min-width: ${MediaSize.Tablet}) {
    position: relative;
    height: 100%;
    width: 100%;
    overflow-y: scroll;
    padding: .25rem 6px .25rem 0;
    margin: 0;
  }
`;

class DetailView extends Component {
  componentWillMount() {
    const { labels, faces, getImage, image: { id } } = this.props;

    // request labels and faces from backend
    if (labels.length === 0 && faces.length === 0) {
      getImage(id);
    }
  }

  render() {
    const { labels, faces, image, getImageRequest } = this.props;
    const { loading } = getImageRequest;
    console.log(this.props);
    return (
      <View>
        <StyledImageBox>
          <AsyncImage src={getImageSrc(image.path)} fit="contain" />
        </StyledImageBox>
        <StyledDataBox pad="small">
          <StyledScrollableData>
            <StyledHeading style={{ marginTop: 0 }}level="4">Labels ({labels.length})</StyledHeading>
            { loading ? (
              <LoadingIndicator />
            ) : (
              <Labels
                labels={labels}
                onLabelClick={(label) => console.log(label)}
              />
            )}
            <StyledHeading level="4">Faces ({faces.length})</StyledHeading>
            { loading ? (
              <LoadingIndicator />
            ) : (
              <Faces
                faces={faces}
                onFaceClick={(face) => console.log(face)}
              />
            )}
          </StyledScrollableData>
        </StyledDataBox>
      </View>
    );
  }
}

DetailView.propTypes = {
  image: PropTypes.shape({
    id: PropTypes.string.isRequired,
    path: PropTypes.string.isRequired,
    name: PropTypes.string,
    type: PropTypes.string,
  }).isRequired,
  labels: PropTypes.arrayOf(PropTypes.shape({})).isRequired,
  faces: PropTypes.arrayOf(PropTypes.shape({})).isRequired,
  getImageRequest: HOCRequestPropTypes.isRequired,
  getImage: PropTypes.func.isRequired,
}

DetailView.defaultProps = {

}

export default DetailView;

