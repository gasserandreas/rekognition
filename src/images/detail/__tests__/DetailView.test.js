/* globals testUtils */
import React from 'react';
import { mount, shallow } from 'enzyme';
import { Provider } from 'react-redux';
import toJson from 'enzyme-to-json';
import 'jest-styled-components';

import DetailView, { __testables__ } from '../DetailView';
import Labels from '../Labels';
import Label from '../Label';
import Faces, { Face } from '../Faces';
import { Attribute } from '../Attribute';
import Image from '../Image';
import AddImageButton from '../../AddImageButton';

import * as Paths from '../../../paths';

import imageJson from '../__data__/image.json';
import labelsJson from '../__data__/labels.json';
import facesJson from '../__data__/faces.json';
import selectedFaceJson from '../__data__/selectedFace.json';
import selectedLabelJson from '../__data__/selectedLabel.json';

const {
  StyledBackButton,
  StyledHeading,
  StyledImageBoxContainer,
  StyledImageBox,
  StyledDataBox,
  StyledScrollableData,
  StyledView,
  convertMetaToAttributes,
} = __testables__;

const { createMockStore } = testUtils;

describe('DetailView test suite', () => {
  const initialProps = {
    history: {
      push: jest.fn(),
    },
    image: imageJson,
    faces: facesJson,
    labels: labelsJson,
    selectedFace: selectedFaceJson,
    selectedLabel: selectedLabelJson,
    getImage: jest.fn(),
    getImageRequest: {
      error: null,
      loading: false,
    },
  };

  const mockStore = createMockStore();

  afterEach(() => {
    // clear mocks
    initialProps.history.push.mockClear();
    initialProps.getImage.mockClear();
  });

  const getDetailView = (props) => {
    const store = mockStore();
    const provider = mount(
      <Provider store={store}>
        <DetailView {...initialProps} {...props} />
      </Provider>,
    );

    const wrapper = provider.find(DetailView);
    return wrapper;
  };

  it('should render', () => {
    const wrapper = getDetailView();
    expect(wrapper.exists()).toBeTruthy();
  });

  it('should render AddImageButton', () => {
    const wrapper = getDetailView();
    expect(wrapper.exists()).toBeTruthy();

    const addImageButton = wrapper.find(AddImageButton);
    expect(addImageButton.exists()).toBeTruthy();
  });

  it('should not render Image component if no image id is provided', () => {
    const wrapper = getDetailView({
      image: {},
    });
    expect(wrapper.exists()).toBeTruthy();

    const image = wrapper.find(Image);
    expect(image.exists()).toBeFalsy();
  });

  it('should render image with props', () => {
    const wrapper = getDetailView();
    expect(wrapper.exists()).toBeTruthy();

    const image = wrapper.find(Image);
    expect(image.exists()).toBeTruthy();

    const props = image.props();
    expect(props.image).toEqual(initialProps.image);
    expect(props.selectedLabel).toEqual(initialProps.selectedLabel);
    expect(props.selectedFace).toEqual(initialProps.selectedFace);
  });

  it('should render back button', () => {
    const wrapper = getDetailView();
    expect(wrapper.exists()).toBeTruthy();

    const backButton = wrapper.find(StyledBackButton);
    expect(backButton.exists()).toBeTruthy();
  });

  it('should render image attributes', () => {
    const wrapper = getDetailView();
    expect(wrapper.exists()).toBeTruthy();

    const imageAttributesWrapper = wrapper.find('#jestImageAttributes');
    expect(imageAttributesWrapper.exists()).toBeTruthy();

    const metaAttributes = convertMetaToAttributes(initialProps.image.meta);

    const attributes = imageAttributesWrapper.find(Attribute);
    expect(attributes.length).toEqual(metaAttributes.length + 1);

    const attributeNames = ['Uploaded', 'Type', 'Size', 'Dimension', 'Density'];
    attributes.forEach((attribute) => {
      expect(attributeNames.includes(attribute.props().attribute.name)).toBeTruthy();
    });
  });

  it('should render labels if not loading', () => {
    const wrapper = getDetailView({
      ...initialProps,
      getImageRequest: {
        ...initialProps.getImageRequest,
        loading: false,
      },
    });
    expect(wrapper.exists()).toBeTruthy();

    const labels = wrapper.find(Labels);
    expect(labels.exists()).toBeTruthy();

    const props = labels.props();
    expect(props.labels).toEqual(initialProps.labels);
    expect(props.selectedLabel).toEqual(initialProps.selectedLabel);
  });

  it('should render loading state for labels if loading', () => {
    const wrapper = getDetailView({
      ...initialProps,
      getImageRequest: {
        ...initialProps.getImageRequest,
        loading: true,
      },
    });
    expect(wrapper.exists()).toBeTruthy();

    const asyncContainer = wrapper.find('#jestLabelsAsyncContainer');
    expect(asyncContainer.exists()).toBeTruthy();

    const labels = wrapper.find(Labels);
    expect(labels.exists()).toBeFalsy();
  });

  it('should not load faces if number of faces is equal zero', () => {
    const wrapper = getDetailView({
      ...initialProps,
      faces: [],
    });
    expect(wrapper.exists()).toBeTruthy();

    const faces = wrapper.find('#jestFacesContainer');
    expect(faces.exists()).toBeFalsy();
  });

  it('should render faces if not loading', () => {
    const wrapper = getDetailView({
      ...initialProps,
      getImageRequest: {
        ...initialProps.getImageRequest,
        loading: false,
      },
    });
    expect(wrapper.exists()).toBeTruthy();

    const faces = wrapper.find(Faces);
    expect(faces.exists()).toBeTruthy();

    const props = faces.props();
    expect(props.faces).toEqual(initialProps.faces);
    expect(props.selectedFace).toEqual(initialProps.selectedFace);
  });

  it('should render loading state for faces if loading', () => {
    const wrapper = getDetailView({
      ...initialProps,
      getImageRequest: {
        ...initialProps.getImageRequest,
        loading: true,
      },
    });
    expect(wrapper.exists()).toBeTruthy();

    const asyncContainer = wrapper.find('#jestFacesAsyncContainer');
    expect(asyncContainer.exists()).toBeTruthy();

    const faces = wrapper.find(Faces);
    expect(faces.exists()).toBeFalsy();
  });

  it('should request image details if labels and faces are not set', () => {
    expect(initialProps.getImage).not.toHaveBeenCalled();

    const wrapper = getDetailView({
      ...initialProps,
      faces: [],
      labels: [],
    });
    expect(wrapper.exists()).toBeTruthy();

    expect(initialProps.getImage).toHaveBeenCalled();
    expect(initialProps.getImage).toHaveBeenCalledWith(initialProps.image.id);
  });

  it('should move to home view after new image upload', () => {
    const wrapper = getDetailView();
    expect(wrapper.exists()).toBeTruthy();

    const addImageButton = wrapper.find(AddImageButton);
    expect(addImageButton.exists()).toBeTruthy();

    expect(initialProps.history.push).not.toHaveBeenCalled();

    addImageButton.props().afterOnClick();
    expect(initialProps.history.push).toHaveBeenCalled();
  });

  it('should convert meta values to attributes', () => {
    const { meta } = initialProps.image;

    const attributes = convertMetaToAttributes(meta);
    expect(attributes).toEqual([
      {
        name: 'Type',
        value: 'jpeg',
      },
      {
        name: 'Size',
        value: '2.26 MB',
      },
      {
        name: 'Dimension',
        value: '4813px x 3209px',
      },
      {
        name: 'Density',
        value: '72 DPI',
      },
    ]);
  });

  it('should return empty array if no meta is provided', () => {
    expect(convertMetaToAttributes(undefined)).toEqual([]);
  });

  it('should filter out invalid image meta properties', () => {
    const meta = {
      ...initialProps.image.meta,
      size: 0,
      height: 0,
      density: 0,
    };

    const attributes = convertMetaToAttributes(meta);
    expect(attributes).toEqual([
      {
        name: 'Type',
        value: 'jpeg',
      },
    ]);
  });

  it('should pass face id to browser on face click', () => {
    const wrapper = getDetailView();
    expect(wrapper.exists()).toBeTruthy();

    const faces = wrapper.find(Face);
    expect(faces.exists()).toBeTruthy();

    expect(initialProps.history.push).not.toHaveBeenCalled();

    // click one face
    const face = faces.at(0);
    expect(face.exists()).toBeTruthy();

    face.simulate('click');

    expect(initialProps.history.push).toHaveBeenCalled();
  });

  it('should pass label id to browser on label click', () => {
    const wrapper = getDetailView();
    expect(wrapper.exists()).toBeTruthy();

    const labels = wrapper.find(Label);
    expect(labels.exists()).toBeTruthy();

    expect(initialProps.history.push).not.toHaveBeenCalled();

    // click one face
    const label = labels.at(2);
    expect(label.exists()).toBeTruthy();

    label.simulate('click');

    expect(initialProps.history.push).toHaveBeenCalled();
  });

  it('should go back on back button click', () => {
    const wrapper = getDetailView();
    expect(wrapper.exists()).toBeTruthy();

    const backButton = wrapper.find(StyledBackButton);
    expect(backButton.exists()).toBeTruthy();

    expect(initialProps.history.push).not.toHaveBeenCalled();

    backButton.simulate('click');
    expect(initialProps.history.push).toHaveBeenCalled();
    expect(initialProps.history.push).toHaveBeenCalledWith(Paths.HOME);
  });

  // check for styling
  it('should render StyledBackButton consistently', () => {
    const wrapper = shallow(<StyledBackButton />);
    expect(toJson(wrapper.dive())).toMatchSnapshot();
  });

  it('should render StyledHeading consistently', () => {
    const wrapper = shallow(<StyledHeading />);
    expect(toJson(wrapper.dive())).toMatchSnapshot();
  });

  it('should render StyledImageBoxContainer consistently', () => {
    const wrapper = shallow(<StyledImageBoxContainer />);
    expect(toJson(wrapper.dive())).toMatchSnapshot();
  });

  it('should render StyledImageBox consistently', () => {
    const wrapper = shallow(<StyledImageBox />);
    expect(toJson(wrapper.dive())).toMatchSnapshot();
  });

  it('should render StyledDataBox consistently', () => {
    const wrapper = shallow(<StyledDataBox />);
    expect(toJson(wrapper.dive())).toMatchSnapshot();
  });

  it('should render StyledScrollableData consistently', () => {
    const wrapper = shallow(<StyledScrollableData />);
    expect(toJson(wrapper.dive())).toMatchSnapshot();
  });

  it('should render StyledView consistently', () => {
    const wrapper = shallow(<StyledView />);
    expect(toJson(wrapper.dive())).toMatchSnapshot();
  });
});
