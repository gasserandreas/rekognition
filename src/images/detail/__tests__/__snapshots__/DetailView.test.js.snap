// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DetailView test suite should render StyledBackButton consistently 1`] = `
<ContextConsumer>
  <Component />
</ContextConsumer>
`;

exports[`DetailView test suite should render StyledDataBox consistently 1`] = `
<ContextConsumer>
  <Component />
</ContextConsumer>
`;

exports[`DetailView test suite should render StyledHeading consistently 1`] = `
<ContextConsumer>
  <Component />
</ContextConsumer>
`;

exports[`DetailView test suite should render StyledImageBox consistently 1`] = `
<ContextConsumer>
  <Component />
</ContextConsumer>
`;

exports[`DetailView test suite should render StyledImageBoxContainer consistently 1`] = `
<ContextConsumer>
  <Component />
</ContextConsumer>
`;

exports[`DetailView test suite should render StyledScrollableData consistently 1`] = `
<ContextConsumer>
  <Component />
</ContextConsumer>
`;

exports[`DetailView test suite should render StyledView consistently 1`] = `
<ContextConsumer>
  <Component />
</ContextConsumer>
`;
